### PR TITLE
Add router with the `proxy` service for rolling node

### LINKED ISSUE #58: Add router with the `proxy` service for rolling node
Description:
- Add router handling in the [usecase](https://github.com/octo-technology/tezos-link/blob/7ee53ed654cc031385448e01f6254165b7c002ad/internal/proxy/usecases/proxy_usecase.go#L53)
- Add to the proxy config 
   - The endpoints : `ARCHIVE_NODES_URL` and `ROLLING_NODES_URL`
   - The path to redirect. We'll start with all `/blocks/heads` redirected to the `ROLLING_NODES_URL`

Estimation: 3

Associated MR:
- https://github.com/octo-technology/tezos-link/pull/50

### PR DESCRIPTION
**Pull request by frankhillard**
Question: should IsRollingRedirection be in the BlockchainRepository interface ?
proxyBlockchainRepository -> implemented
lruBlockchainRepository -> not implemented

### CODE REVIEW

File: config/proxy_config.go [hunk 1]
```diff
@@ -14,8 +14,11 @@ import (
 type ProxyConf struct {
 	Debug bool
 	Tezos struct {
-		Host string
-		Port int
+		ArchiveHost string
+		Port        int
```
Comment (louptheron) [inline config/proxy_config.go]: ArchivePort ?

File: config/proxy_config.go [hunk 2]
```diff
@@ -57,12 +60,18 @@ func ParseProxyConf(cfg string) (*ProxyConf, error) {
...
+	conf.Tezos.RollingHost = getEnv("ROLLING_NODES_URL", "node")
 	tezosPort, err := strconv.Atoi(getEnv("TEZOS_PORT", "1090"))
 	if err != nil {
 		logrus.Fatal(err)
 	}
+	tezosRollingPort, err := strconv.Atoi(getEnv("TEZOS_ROLLING_PORT", "1090"))
+	if err != nil {
+		logrus.Fatal(err)
+	}
 	conf.Tezos.Port = tezosPort
```
Comment (louptheron) [inline config/proxy_config.go]: ArchivePort ?

File: data/proxy/conf/local.toml [hunk 3]
```diff
@@ -1,5 +1,10 @@
 debug = true
 
+[tezos]
+whitelistedRolling = [
```
Comment (louptheron) [inline data/proxy/conf/local.toml]: To be consistent, could be rename whitelisted to whitelistedArchive ? And same for the other properties

File: internal/proxy/infrastructure/cache/lru_blockchain_repository.go [hunk 4]
```diff
@@ -39,3 +39,7 @@ func (l lruBlockchainRepository) Add(request *pkgmodel.Request, response interfa
 
 	return nil
 }
+
+func (p lruBlockchainRepository) IsRollingRedirection(url string) bool {
```
Comment (louptheron) [inline internal/proxy/infrastructure/cache/lru_blockchain_repository.go]: IsRollingRedirection is more a business logic than an interaction with the blockchain repository, so I suggest to move this method to the proxy use case

**Review by louptheron — CHANGES_REQUESTED** [review]
I added few comments.
Also, I suggest to move the IsRollingRedirection method and related tests to the proxy use case, as the is no blockchain repository involved but only some business rules from config files

**Comment by aboulay** [comment]
I wonder if we could rename environment variables to get something more constant :

`ARCHIVE_NODES_URL` -> ✅  OK
`TEZOS_PORT` -> `ARCHIVE_NODES_PORT`
`ROLLING_NODES_URL` -> ✅  OK
`TEZOS_ROLLING_PORT` -> `ROLLING_NODES_PORT`

File: data/proxy/conf/local.toml (hunk 3 above)
Comment (frankhillard) [reply data/proxy/conf/local.toml]: Disagree on that. The goal is to redirect the request  .... by default we redirect to archive nodes. whitelistedRolling represents patterns of url for which will redirect to rolling nodes. Maybe we can rename whitelistedRolling into patternsRolling ?

File: config/proxy_config.go [hunk 5]
```diff
@@ -34,6 +36,7 @@ type ProxyConf struct {
 		RateLimitCount                  int64
 		BlockchainRequestsCacheMaxItems int
 		ProjectsCacheMaxItems           int
+		WhitelistedRolling	            []string
```
Comment (aboulay) [inline config/proxy_config.go:42]: Can you apply a `fmt` please ? :)

File: internal/proxy/usecases/helpers_test.go [hunk 6]
```diff
@@ -22,6 +22,11 @@ func (m *mockBlockchainRepository) Add(request *pkgmodel.Request, response inter
 	return nil
 }
 
+//func (m *mockBlockchainRepository) IsRollingRedirection(url string) bool {
```
Comment (aboulay) [inline internal/proxy/usecases/helpers_test.go]: deadcode ?

**Review by aboulay — CHANGES_REQUESTED** [review]
You get my review. This is nice !
Just few changes and we are done. o/

**Review by aboulay — APPROVED** [review]

**Review by louptheron — APPROVED** [review]

**Pull request merged**